akka-typed/src/main/scala/akka/typed/javadsl/BehaviorChain.scala

```diff
@@ -0,0 +1,243 @@
+/**
+ * Copyright (C) 2009-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.typed.javadsl
+
+import scala.annotation.tailrec
+import akka.japi.function.{ Function, Function2, Predicate }
+import akka.typed
+import akka.typed.{ Behavior, ExtensibleBehavior, Signal }
+import BehaviorChain._
+import akka.annotation.InternalApi
+
+/**
+ * Used for creating a [[Behavior]] by 'chaining' message and signal handlers.
+ *
+ * When handling a message or signal, this [[Behavior]] will consider all handlers in the order they were added,
+ * looking for the first handler for which both the type and the (optional) predicate match.
+ *
+ * @tparam T the common superclass of all supported messages.
+ */
+class BehaviorChain[T] private (
```

Comment:
What was the rationale behind not reusing the existing `akka.japi.JavaPartialFunction` which is on top of the Scala partial function impl?